File: src/ijp_tgtnum/ijp.py

tmp = open("test.h", "w")
for i in range(0,512):
    tmp.write("void foo_func"+str(i)+"(){;}\n")
tmp.write("\n")

tmp.write(" #define ASSIGN ")
for i in range(0,512):
    tmp.write(f"tgt[{i}] = foo_func{i};\\\n")
tmp.write("\n")

tmp.write(" #define TWO ")
for i in range(0,2):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define FOUR ")
for i in range(0,4):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define EIGHT ")
for i in range(0,8):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define SIXTEEN ")
for i in range(0,16):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define THIRTYTWO ")
for i in range(0,32):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define SIXTYFOUR ")
for i in range(0,64):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define HUNDRED ")
for i in range(0,128):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define THUNDRED ")
for i in range(0,256):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

tmp.write(" #define FHUNDRED ")
for i in range(0,512):
    if(i  == 0):
        tmp.write(f"a[{i}] = tgt[pattern[i]]; a[{i}](); \\\n")
    else:
        tmp.write(f"a[{i}] = tgt[pattern[(i+16*{i})%MAXHIS]]; a[{i}](); \\\n")
tmp.write("\n")

File: src/ijp_tgtnum_new/ijp.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <sys/time.h>
#include "test.h"

#define MAXHIS 4096
#define MAXTGT 512
#define MAXNUM 512
#define REPEAT 100

int mygen(unsigned int*dst,uint64_t pattern){
    for(uint64_t i=0; i<MAXHIS; i++){
        dst[i] = i%pattern;
    }
    return 0;
}


int main(){
    void (* a[MAXNUM])();
    void (* tgt[MAXTGT])();
    int b=0;
    ASSIGN

    struct timeval tv_begin, tv_end;
    unsigned int pattern[MAXHIS];
    int tgt_num[] = {2,4,8,12,16,24,32,48,64,96,128,192,256,384,512};

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT*4;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                a[0] = tgt[pattern[i]]; a[0]();
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*4));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT*4;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                TWO
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*4*2));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT*4;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                FOUR
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*4*4));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                EIGHT
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*8));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                SIXTEEN
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*16));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                THIRTYTWO
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*32));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                SIXTYFOUR
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*64));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                HUNDRED
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*128));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                THUNDRED
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*256));
    }
    printf("\n");

    for(int i=0;i<15;i++){ 
        mygen(pattern,tgt_num[i]);
        gettimeofday(&tv_begin, NULL);
        for(uint64_t j=0;j<REPEAT;j++){
            for(uint64_t i=0; i<MAXHIS;i++){
                FHUNDRED
            }
        }
        gettimeofday(&tv_end, NULL);
        double timeSpan = (tv_end.tv_sec - tv_begin.tv_sec)* 1000000 + (tv_end.tv_usec-tv_begin.tv_usec);
        printf("%f,",(timeSpan*1000.0)/(MAXHIS*REPEAT*512));
    }
    printf("\n");

    return b;
}
